### ISSUE TITLE
[Sulong] "VMError.shouldNotReachHere: Attempting to do a Native-to-Java transition when already in Java mode" when trying to execute Clang 7

### ISSUE THREAD

**Issue by Godin** [oracle/graal#901]
Building of Clang into bitcode is a bit tricky, but possible as following.

Main idea is to use Gold Linker:

```
$ cat <<END > test.cpp
int main() {
  return 42;
}
END

$ clang++ -fuse-ld=gold -flto -Wl,-plugin-opt=save-temps test.cpp

$ ~/graalvm-ce-1.0.0-rc10/bin/lli a.out.0.0.preopt.bc

$ echo $?
42
```

For build of Clang I used `n1-standard-32` machine in Google Compute Engine with `ubuntu-1804-lts` image and following configuration after installation of CMake, Ninja, Clang 6.0, Gold Linker, etc:

```
$ git clone https://github.com/llvm-mirror/llvm.git --branch=release_70 --single-branch

$ git clone https://github.com/llvm-mirror/clang.git --branch=release_70 --single-branch llvm/tools/clang

$ mkdir build && cd build

$ CXX=clang++ CC=clang \
cmake \
  -GNinja \
  -DCMAKE_BUILD_TYPE=RELEASE \
  -DLLVM_ENABLE_TERMINFO=OFF \
  -DLLVM_ENABLE_THREADS=OFF \
  -DLLVM_ENABLE_ZLIB=OFF \
  -DLLVM_TARGETS_TO_BUILD=X86 \
  -LLVM_USE_LINKER=gold \
  -DLLVM_ENABLE_LTO=ON \
  -DCMAKE_RANLIB=/usr/bin/llvm-ranlib-6.0 \
  -DCMAKE_AR=/usr/bin/llvm-ar-6.0 \
  -DCMAKE_CXX_FLAGS="--stdlib=libc++" \
  -DCMAKE_EXE_LINKER_FLAGS="-fuse-ld=gold -flto -Wl,-plugin-opt=save-temps" \
  ../llvm

$ ninja
```

after build execution of

```
$ ~/graalvm-ce-1.0.0-rc10/bin/lli ./bin/clang-7.0.0.preopt.bc --version
```

fails with

```
VMError.shouldNotReachHere: Attempting to do a Native-to-Java transition when already in Java mode
```

full [log.txt](https://github.com/oracle/graal/files/2741096/log.txt) ,

and execution of

```
$ ~/graalvm-ce-1.0.0-rc10/bin/lli --jvm ./bin/clang-7.0.0.preopt.bc --version
```

fails with

```
Segmentation fault (core dumped)
```

Same happens with build of Sulong from [ed108a72b8fa94374a12057d9d0e16173be21a2a](https://github.com/oracle/graal/commit/ed108a72b8fa94374a12057d9d0e16173be21a2a).

Another binary from LLVM partially works:

```
$ ./bin/not
$ echo $?
1

$ ./bin/not /bin/false
$ echo $?
0

$ ./bin/not x
Error: Unable to find `x' in PATH: No such file or directory

$ ~/graalvm-ce-1.0.0-rc10/bin/lli ./bin/not.0.0.preopt.bc
$ echo $?
1

$ ~/graalvm-ce-1.0.0-rc10/bin/lli ./bin/not.0.0.preopt.bc /bin/false
ERROR: com.oracle.truffle.api.dsl.UnsupportedSpecializationException: Unexpected values provided for LLVMSignedCastToI64VectorNodeGen@4cc74724: [< 2 x void* > < void* 0x4b6c8e0, void* 0x4b6c8e0 >], [LLVMPointerVector]

$ ~/graalvm-ce-1.0.0-rc10/bin/lli ./bin/not.0.0.preopt.bc x
Error: Unable to find `x' in PATH: No such file or directory
```

**Comment by Godin** [comment]
FWIW same still happens with RC15:

```
$ ~/graalvm-ce-1.0.0-rc15/bin/lli ./bin/clang-8.0.0.preopt.bc --version
Fatal error: Attempting to do a Native-to-Java transition when already in Java mode

$ ~/graalvm-ce-1.0.0-rc15/bin/lli --jvm ./bin/clang-8.0.0.preopt.bc --version
Segmentation fault (core dumped)
```

**Comment by christianwimmer** [comment]
@Godin Can you share the complete diagnostic report of the Substrate VM "Fatal error", and the crash log file of HotSpot when running in `--jvm` mode?

**Comment by Godin** [comment]
@christianwimmer Thank you for looking into this problem. Here is [log.txt](https://github.com/oracle/graal/files/3069338/log.txt) of `~/graalvm-ce-1.0.0-rc15/bin/lli ./bin/clang-8.0.0.preopt.bc --version` failure. For some reasons out of ~20 executions with `--jvm` I only once got usual crash message [log2.txt](https://github.com/oracle/graal/files/3069476/log2.txt) and [hs_err_pid2551.log](https://github.com/oracle/graal/files/3069477/hs_err_pid2551.log) without any `replay_*` file, and for all others there is only message "Segmentation fault".

@shelajev had an idea to create docker image as a reproducer. WDYT - can this help you with debugging?

**Comment by christianwimmer** [comment]
@Godin thanks for the log files. They are interesting - in the HotSpot crash it seems like application code gets executed on a JIT compiler thread. Looks like some bad memory corruption. I'll try to reproduce on native image, since that seems to be a good path to debugging the issue.

Can you share the file `clang-8.0.0.preopt.bc` so that I don't need to build it myself?

**Comment by Godin** [comment]
@christianwimmer surprised that only this file will be enough, given that bitcode is platform dependent, and I guess Oleg was suggesting Docker image due to this. Nevertheless uploaded  to Dropbox - https://www.dropbox.com/s/tasnsxi3uludpv8/clang-8.0.0.preopt.bc.tar.gz ( it doesn't fit into GitHub limit for attachments even after compression 😄 ).

**Comment by rschatz** [comment]
Should be fixed.

The fix is already in the latest release (19.3), sorry for not closing this earlier. It was fixed implicitly by some other issue and we missed that one.
